cpp/src/fil/fil.cu

```diff
@@ -71,43 +71,63 @@ __global__ void transform_k(float* preds, size_t n, output_t output,
 }
 
 struct forest {
-  void init_max_shm() {
+  void enumerate_n_items() {
+    for (ssp_.n_items = algo_ == algo_t::BATCH_TREE_REORG ? 4 : 1;
```

Comment:
This loop looks unnecessarily complex. Consider making `n_items` a local variable.